File: December2021/C++/house_robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size()<2) 
            return nums[0];
        vector<int> dp(nums.size(),-1);
        return max(knapsack(nums,nums.size()-1,dp),knapsack(nums,nums.size()-2,dp));
    }
    
    int knapsack(vector<int>& nums,int n,vector<int>& dp)
    {
        if(n>=0 && dp[n]!=-1)
            return dp[n];
        if(n>=0)
        {
            dp[n]=max(knapsack(nums,n-2,dp)+nums[n],knapsack(nums,n-1,dp));
            return dp[n];
        }
        return 0;
    }
};

File: December2021/HouseRobber.java
class Solution {
	// TC : O(n)

	public int rob(int[] nums) {
		if(nums.length==0) {
			return 0;
		} else if(nums.length==1) {
			return nums[0];
		}
		else if(nums.length==2) {
			return Math.max(nums[0], nums[1]);
		}
		else {

			int[] dp = new int[nums.length];
			dp[0]=nums[0];
			dp[1]= nums[1];
			dp[2]= nums[0]+nums[2];
			int max = Math.max(dp[2], dp[1]);
			for(int i=3;i<nums.length;i++){

				dp[i]=  Math.max(dp[i-2], dp[i-3]) + nums[i];
				max = Math.max(dp[i], max);
			}


			return max;
		}
	}
}